Parsing error on parenthesised anonymous procedure

### Prerequisites

- [X] This bug is in SonarDelphi, not SonarQube or my Delphi code.
- [X] This bug has not already been reported.

### SonarDelphi version

1.4.0

### SonarQube version

_No response_

### Issue description

When an anonymous procedure literal is called directly, sonar-delphi fails to parse it.

### Steps to reproduce

Scan the following delphi program

```delphi
program Example;
begin
  (procedure(a: String) begin end)('');
end.
```

and observe the error

```
au.com.integradev.delphi.file.DelphiFile$DelphiFileConstructionException: Failed to construct DelphiFile (line 4:3 no viable alternative at input 'procedure')
	at au.com.integradev.delphi.file.DelphiFile.setupFile(DelphiFile.java:159)
	at au.com.integradev.delphi.file.DelphiFile$DelphiInputFile.from(DelphiFile.java:75)
	at au.com.integradev.delphi.DelphiSensor.executeOnFiles(DelphiSensor.java:143)
	at au.com.integradev.delphi.DelphiSensor.execute(DelphiSensor.java:83)
	at org.sonarsource.sonarlint.core.analysis.container.analysis.sensor.SensorsExecutor.executeSensor(SensorsExecutor.java:75)
	at org.sonarsource.sonarlint.core.analysis.container.analysis.sensor.SensorsExecutor.execute(SensorsExecutor.java:66)
	at org.sonarsource.sonarlint.core.analysis.container.analysis.AnalysisContainer.doAfterStart(AnalysisContainer.java:122)
	at org.sonarsource.sonarlint.core.plugin.commons.container.SpringComponentContainer.startComponents(SpringComponentContainer.java:182)
	at org.sonarsource.sonarlint.core.plugin.commons.container.SpringComponentContainer.execute(SpringComponentContainer.java:161)
	at org.sonarsource.sonarlint.core.analysis.container.module.ModuleContainer.analyze(ModuleContainer.java:71)
	at au.com.integradev.delphilint.analysis.AnalysisOrchestrator.runAnalysis(AnalysisOrchestrator.java:147)
	at au.com.integradev.delphilint.server.AnalysisServer.analyze(AnalysisServer.java:115)
	at au.com.integradev.delphilint.server.TlvConnection.processRequest(TlvConnection.java:171)
	at au.com.integradev.delphilint.server.TlvConnection.readStream(TlvConnection.java:158)
	at au.com.integradev.delphilint.server.TlvConnection.run(TlvConnection.java:70)
	at au.com.integradev.delphilint.App.main(App.java:70)
Caused by: au.com.integradev.delphi.antlr.DelphiParser$ParserException: line 4:3 no viable alternative at input 'procedure'
	at au.com.integradev.delphi.antlr.DelphiParser.reportError(DelphiParser.java:386)
	at au.com.integradev.delphi.antlr.DelphiParser.statement(DelphiParser.java:25090)
	at au.com.integradev.delphi.antlr.DelphiParser.delimitedStatements(DelphiParser.java:29460)
	at au.com.integradev.delphi.antlr.DelphiParser.statementList(DelphiParser.java:29345)
	at au.com.integradev.delphi.antlr.DelphiParser.compoundStatement(DelphiParser.java:29273)
	at au.com.integradev.delphi.antlr.DelphiParser.programBody(DelphiParser.java:1121)
	at au.com.integradev.delphi.antlr.DelphiParser.program(DelphiParser.java:796)
	at au.com.integradev.delphi.antlr.DelphiParser.file(DelphiParser.java:495)
	at au.com.integradev.delphi.file.DelphiFile.createAST(DelphiFile.java:194)
	at au.com.integradev.delphi.file.DelphiFile.setupFile(DelphiFile.java:147)
	... 15 more
Caused by: NoViableAltException(129@[])
	at au.com.integradev.delphi.antlr.DelphiParser.statement(DelphiParser.java:24809)
	... 23 more
```

### Minimal Delphi code exhibiting the issue

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,6 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - Symbol table construction errors for program files omitting the `program` header.
+- Parsing errors on parenthesised anonymous methods.
 - Incorrect ordering of edits in the "Separate grouped parameters" quick fix for `GroupedParameterDeclaration`.
 
 ## [1.5.0] - 2024-05-02

--- a/delphi-frontend/src/main/antlr3/au/com/integradev/delphi/antlr/Delphi.g
+++ b/delphi-frontend/src/main/antlr3/au/com/integradev/delphi/antlr/Delphi.g
@@ -697,6 +697,7 @@ attribute                    : (ASSEMBLY ':')? nameReference argumentList? (':' 
 // Expressions
 //----------------------------------------------------------------------------
 expression                   : relationalExpression
+                             | anonymousMethod
                              ;
 // ANTLR sets the begin and end tokens for nested binary expression nodes
 // in relationalOperator, not relationalExpression, meaning that their
@@ -748,8 +749,7 @@ argument                     : argumentName? argumentExpression
 argumentName                 : ident ':='!
                              | keywords ':=' -> ^({changeTokenType(TkIdentifier, -2)})
                              ;
-argumentExpression           : anonymousMethod
-                             | expression writeArguments?
+argumentExpression           : expression writeArguments?
                              ;
 writeArguments               : ':'! expression (':'! expression)? // See: https://docwiki.embarcadero.com/Libraries/en/System.Write
                              ;
@@ -758,17 +758,9 @@ anonymousMethod              : PROCEDURE<AnonymousMethodNodeImpl>^ routineParame
                              ;
 expressionOrRange            : expression ('..'<RangeExpressionNodeImpl>^ expression)?
                              ;
-expressionOrAnonymousMethod  : anonymousMethod
-                             | expression
-                             ;
-exprOrRangeOrAnonMethod      : anonymousMethod
-                             | expressionOrRange
-                             ;
 expressionList               : (expression (','!)?)+
                              ;
 expressionOrRangeList        : (expressionOrRange (','!)?)+
-                             ;
-exprOrRangeOrAnonMethodList  : (exprOrRangeOrAnonMethod (','!)?)+
                              ;
 textLiteral                  : singleLineTextLiteral -> ^(TkTextLiteral<TextLiteralNodeImpl> singleLineTextLiteral)
                              | multilineTextLiteral -> ^(TkTextLiteral<TextLiteralNodeImpl> multilineTextLiteral)
@@ -783,8 +775,8 @@ escapedCharacter             : TkCharacterEscapeCode
                              ;
 nilLiteral                   : NIL<NilLiteralNodeImpl>
                              ;
-arrayConstructor             : lbrack exprOrRangeOrAnonMethodList? rbrack
-                             -> ^(TkArrayConstructor<ArrayConstructorNodeImpl> lbrack exprOrRangeOrAnonMethodList? rbrack)
+arrayConstructor             : lbrack expressionOrRangeList? rbrack
+                             -> ^(TkArrayConstructor<ArrayConstructorNodeImpl> lbrack expressionOrRangeList? rbrack)
                              ;
 addOperator                  : '+'<BinaryExpressionNodeImpl>
                              | '-'<BinaryExpressionNodeImpl>
@@ -848,11 +840,11 @@ statement                    : ifStatement
                              ;
 ifStatement                  : IF<IfStatementNodeImpl>^ expression THEN statement? (ELSE statement?)?
                              ;
-varStatement                 : VAR attributeList? nameDeclarationList (':' varType)? (':=' expressionOrAnonymousMethod)?
-                             -> ^(VAR<VarStatementNodeImpl> nameDeclarationList (':' varType)? (':=' expressionOrAnonymousMethod)? attributeList?)
+varStatement                 : VAR attributeList? nameDeclarationList (':' varType)? (':=' expression)?
+                             -> ^(VAR<VarStatementNodeImpl> nameDeclarationList (':' varType)? (':=' expression)? attributeList?)
                              ;
-constStatement               : CONST attributeList? nameDeclaration (':' varType)? '=' expressionOrAnonymousMethod
-                             -> ^(CONST<ConstStatementNodeImpl> nameDeclaration (':' varType)? '=' expressionOrAnonymousMethod attributeList?)
+constStatement               : CONST attributeList? nameDeclaration (':' varType)? '=' expression
+                             -> ^(CONST<ConstStatementNodeImpl> nameDeclaration (':' varType)? '=' expression attributeList?)
                              ;
 caseStatement                : CASE<CaseStatementNodeImpl>^ expression OF caseItem* elseBlock? END
                              ;
@@ -881,7 +873,7 @@ delimitedStatements          : (statement | ';')+
                              ;
 labelStatement               : label ':' statement? -> ^(TkLabelStatement<LabelStatementNodeImpl> label statement?)
                              ;
-assignmentStatement          : expression ':='<AssignmentStatementNodeImpl>^ expressionOrAnonymousMethod
+assignmentStatement          : expression ':='<AssignmentStatementNodeImpl>^ expression
                              ;
 expressionStatement          : expression -> ^(TkExpressionStatement<ExpressionStatementNodeImpl> expression)
                              ;

--- a/delphi-frontend/src/test/resources/au/com/integradev/delphi/grammar/AnonymousMethods.pas
+++ b/delphi-frontend/src/test/resources/au/com/integradev/delphi/grammar/AnonymousMethods.pas
@@ -24,4 +24,12 @@ begin
     end);
 end;
 
+function IsStringEmptyImmediateInvocation(const aValue : string) : Boolean;
+begin
+  Result := (function(const aValue : string) : Boolean
+    begin
+      Result := aValue = '';
+    end)(aValue);
+end;
+
 end.

